src/backend/tcop/zenith_wal_redo.c

```diff
@@ -88,11 +102,73 @@ static void PushPage(StringInfo input_message);
 static void ApplyRecord(StringInfo input_message);
 static bool redo_block_filter(XLogReaderState *record, uint8 block_id);
 static void GetPage(StringInfo input_message);
+int enter_seccomp(void);
 
 static BufferTag target_redo_tag;
 
 #define TRACE DEBUG5
 
+/* A quick and dirty toolkit for hand-building a
+   BPF program that can filter syscalls. */
+
+#define ARCH_NR AUDIT_ARCH_X86_64
+
+#define syscall_nr (offsetof(struct seccomp_data, nr))
+#define arch_nr (offsetof(struct seccomp_data, arch))
+
+#define VALIDATE_ARCHITECTURE \
+	BPF_STMT(BPF_LD+BPF_W+BPF_ABS, arch_nr), \
+	BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, ARCH_NR, 1, 0), \
+	BPF_STMT(BPF_RET+BPF_K, SECCOMP_RET_KILL)
+
+#define EXAMINE_SYSCALL \
+	BPF_STMT(BPF_LD+BPF_W+BPF_ABS, syscall_nr)
+
+#define ALLOW_SYSCALL(name) \
+	BPF_JUMP(BPF_JMP+BPF_JEQ+BPF_K, __NR_##name, 0, 1), \
+	BPF_STMT(BPF_RET+BPF_K, SECCOMP_RET_ALLOW)
+
+#define KILL_PROCESS \
+	BPF_STMT(BPF_RET+BPF_K, SECCOMP_RET_KILL)
+
+/* enter seccomp mode with a BPF filter that will only allow
+   certain syscalls to proceed */
+#if USE_SECCOMP
+int
+enter_seccomp(void)
+{
+	struct sock_filter filter[] = {
+		/* Validate architecture. */
+		VALIDATE_ARCHITECTURE,
+		/* Grab the system call number. */
+		EXAMINE_SYSCALL,
+		/* List allowed syscalls. */
+		ALLOW_SYSCALL(exit_group),
+		ALLOW_SYSCALL(exit),
+		ALLOW_SYSCALL(read),
```

Comment:
Not a review, just wondering. If I understood correctly after a brief reading about `seccomp`, when process makes a transition into this secure state it can do `write() / read()` only to / from already open file descriptors. So in contrast with ordinary Postgres, which opens new descriptors all the time, this one works for us because we do reads / writes only to / from standard input-output, which descriptors are considered to be open since the program start. Am I right?

Reply:
Yes, that's pretty much it.

`SECCOMP_MODE_STRICT` is the mode you describe: only `read` and `write` are allowed.

Here the code does `SECCOMP_MODE_FILTER` which allows us to define a filter script that runs in kernel space to allow the set of syscalls we define.